File: backend/src/services/draftRenderer.ts
import type { DraftTemplate, ProjectArtifacts } from '../types';

/**
 * 初稿模板渲染服务
 *
 * 不依赖任何外部库（无 LaTeX/Markdown 引擎），使用纯字符串模板组装初稿。
 * 设计目标：
 * 1. 任意时候可调用，artifacts 不全时也能给出可读输出
 * 2. 模板异常时统一降级为 Markdown，避免前端拿到空串
 */

interface RenderInput {
  projectName: string;
  discipline: string;
  question: string;
  artifacts: ProjectArtifacts;
}

interface RenderResult {
  text: string;
  template: DraftTemplate;
}

/** 模板白名单 */
const TEMPLATES: ReadonlyArray<DraftTemplate> = ['ctex', 'ieee', 'journal', 'markdown'];

/** 把章节字典拼接为有序列表 */
function pickSections(
  sections: Record<string, string> | undefined,
): Array<{ key: string; value: string }> {
  if (!sections || typeof sections !== 'object') return [];
  const order = [
    'abstract',
    'introduction',
    'method',
    'results',
    'discussion',
    'conclusion',
  ];
  const keys = Object.keys(sections);
  // 先按预设顺序，再补未在预设里的自定义 key
  const sorted = [...order.filter((k) => keys.includes(k))];
  for (const k of keys) {
    if (!sorted.includes(k)) sorted.push(k);
  }
  return sorted.map((k) => ({ key: k, value: sections[k] ?? '' }));
}

/** 把 figures 渲染为占位符列表 */
function pickFigures(figures: ProjectArtifacts['figures']): string[] {
  if (!Array.isArray(figures) || figures.length === 0) return [];
  return figures.map((f, idx) => {
    if (!f || typeof f !== 'object') return `图 ${idx + 1}`;
    const obj = f as Record<string, unknown>;
    const title = typeof obj.title === 'string' ? obj.title : `图 ${idx + 1}`;
    const caption = typeof obj.caption === 'string' ? obj.caption : '';
    return caption ? `${title}\n${caption}` : title;
  });
}

/** 提取参考文献列表 */
function pickReferences(artifacts: ProjectArtifacts): string[] {
  const lit = artifacts.literature;
  if (!lit || typeof lit !== 'object') return [];
  const obj = lit as Record<string, unknown>;
  const refs = obj.references ?? obj.refs;
  if (!Array.isArray(refs)) return [];
  return refs.map((r) => (typeof r === 'string' ? r : JSON.stringify(r)));
}

// ─────────────────────────── 模板字符串 ───────────────────────────

/** CTEX 模板：中文双栏，含摘要/正文/参考文献 */
const CTEX_TEMPLATE = `\\documentclass[twocolumn]{ctexart}
\\usepackage{graphicx}
\\usepackage{geometry}
\\geometry{a4paper, margin=2cm}
\\title{<TITLE>}
\\author{科研自动化流水线}
\\date{\\today}

\\begin{document}
\\twocolumn[
\\maketitle
\\begin{abstract}
<ABSTRACT>
\\end{abstract}
]

\\section{引言}
<INTRODUCTION>

\\section{方法}
<METHOD>

\\section{结果}
<RESULTS>

\\section{讨论}
<DISCUSSION>

\\section{结论}
<CONCLUSION>

<FIGURES_BLOCK>

\\section*{参考文献}
\\begin{thebibliography}{99}
<REFERENCES>
\\end{thebibliography}

\\end{document}
`;

/** IEEE 模板：英文双栏会议 */
const IEEE_TEMPLATE = `\\documentclass[conference]{IEEEtran}
\\usepackage{graphicx}
\\IEEEoverridecommandlockouts

\\title{<TITLE>}
\\author{Research Automation Pipeline\\\\<DISCIPLINE>}

\\begin{document}
\\maketitle

\\begin{abstract}
<ABSTRACT>
\\end{abstract}

\\section{Introduction}
<INTRODUCTION>

\\section{Method}
<METHOD>

\\section{Results}
<RESULTS>

\\section{Discussion}
<DISCUSSION>

\\section{Conclusion}
<CONCLUSION>

<FIGURES_BLOCK>

\\begin{thebibliography}{00}
<REFERENCES>
\\end{thebibliography}

\\end{document}
`;

/** 期刊模板：中文期刊带页眉页脚占位 */
const JOURNAL_TEMPLATE = `\\documentclass[a4paper,11pt]{article}
\\usepackage[UTF8]{ctex}
\\usepackage{fancyhdr}
\\usepackage{geometry}
\\geometry{margin=2.5cm}

\\pagestyle{fancy}
\\fancyhf{}
\\fancyhead[L]{<TITLE>}
\\fancyhead[R]{<DISCIPLINE>}
\\fancyfoot[C]{\\thepage}

\\title{<TITLE>\\thanks{学科: <DISCIPLINE>}}
\\author{科研自动化流水线}
\\date{\\today}

\\begin{document}
\\maketitle

\\begin{abstract}
<ABSTRACT>
\\end{abstract}

\\section{引言}
<INTRODUCTION>

\\section{研究方法}
<METHOD>

\\section{研究结果}
<RESULTS>

\\section{讨论}
<DISCUSSION>

\\section{结论与展望}
<CONCLUSION>

<FIGURES_BLOCK>

\\section*{参考文献}
<REFERENCES>

\\end{document}
`;

/** Markdown 模板：极简 */
const MARKDOWN_TEMPLATE = `# <TITLE>

> 学科: <DISCIPLINE>

## 摘要
<ABSTRACT>

## 引言
<INTRODUCTION>

## 方法
<METHOD>

## 结果
<RESULTS>

## 讨论
<DISCUSSION>

## 结论
<CONCLUSION>

<FIGURES_BLOCK>

## 参考文献
<REFERENCES>
`;

// ─────────────────────────── 模板选择与渲染 ───────────────────────────

function pickTemplate(t: DraftTemplate): string {
  switch (t) {
    case 'ctex':
      return CTEX_TEMPLATE;
    case 'ieee':
      return IEEE_TEMPLATE;
    case 'journal':
      return JOURNAL_TEMPLATE;
    case 'markdown':
      return MARKDOWN_TEMPLATE;
    default:
      return MARKDOWN_TEMPLATE;
  }
}

/** 渲染 markdown 风格图块（latex 风格模板仍可兼容） */
function renderFiguresBlock(figureList: string[], template: DraftTemplate): string {
  if (figureList.length === 0) return '';
  if (template === 'markdown') {
    return figureList
      .map((caption, idx) => `### 图 ${idx + 1}\n\n${caption}\n`)
      .join('\n');
  }
  return figureList
    .map(
      (caption, idx) =>
        `\\begin{figure}[!t]\n\\centering\n\\includegraphics[width=0.45\\textwidth]{figure${idx + 1}.png}\n\\caption{${caption.replace(/\n/g, ' ')}}\n\\end{figure}`,
    )
    .join('\n\n');
}

function renderReferences(refs: string[]): string {
  if (refs.length === 0) return '\\textit{（暂无）}';
  return refs
    .map((r, idx) => {
      if (r.includes('\\bibitem')) return r;
      return `\\bibitem{ref${idx + 1}} ${r}`;
    })
    .join('\n');
}

/** 把 sections 字典转成结构化文本 */
function sectionToText(key: string, value: string): string {
  // 空值占位，避免模板里出现空段
  if (!value || !value.trim()) return '_（待生成）_';
  return value.trim();
}

/**
 * 渲染论文初稿
 * @param artifacts 流水线产物
 * @param template 模板
 * @param meta 项目元数据（标题/学科/问题）
 */
export function renderDraft(
  artifacts: ProjectArtifacts,
  template: DraftTemplate,
  meta: { projectName: string; discipline: string; question: string },
): RenderResult {
  const safeTemplate: DraftTemplate = TEMPLATES.includes(template) ? template : 'markdown';

  try {
    const tpl = pickTemplate(safeTemplate);
    const sections = pickSections(artifacts.paperSections);
    const sectionMap = new Map(sections.map((s) => [s.key, s.value]));
    const figures = pickFigures(artifacts.figures);
    const refs = pickReferences(artifacts);

    const title = meta.projectName || meta.question || '未命名研究';
    const discipline = meta.discipline || '通用';

    const replacements: Array<[string, string]> = [
      ['<TITLE>', escapeTemplate(title)],
      ['<DISCIPLINE>', escapeTemplate(discipline)],
      ['<QUESTION>', escapeTemplate(meta.question || '')],
      ['<ABSTRACT>', sectionToText('abstract', sectionMap.get('abstract') ?? '')],
      [
        '<INTRODUCTION>',
        sectionToText('introduction', sectionMap.get('introduction') ?? ''),
      ],
      ['<METHOD>', sectionToText('method', sectionMap.get('method') ?? '')],
      ['<RESULTS>', sectionToText('results', sectionMap.get('results') ?? '')],
      ['<DISCUSSION>', sectionToText('discussion', sectionMap.get('discussion') ?? '')],
      [
        '<CONCLUSION>',
        sectionToText('conclusion', sectionMap.get('conclusion') ?? ''),
      ],
      ['<FIGURES_BLOCK>', renderFiguresBlock(figures, safeTemplate)],
      ['<REFERENCES>', renderReferences(refs)],
    ];
    let replaced = tpl;
    for (const [token, value] of replacements) {
      replaced = replaceAll(replaced, token, value);
    }

    return { text: replaced, template: safeTemplate };
  } catch (err) {
    // 任何渲染异常都降级为 markdown
    console.warn(
      `[draftRenderer] 渲染失败，降级为 markdown: ${err instanceof Error ? err.message : String(err)}`,
    );
    return {
      text: renderMarkdownFallback(artifacts, meta),
      template: 'markdown',
    };
  }
}

/** 极简 markdown 降级方案 */
function renderMarkdownFallback(
  artifacts: ProjectArtifacts,
  meta: { projectName: string; discipline: string; question: string },
): string {
  const title = meta.projectName || meta.question || '未命名研究';
  const discipline = meta.discipline || '通用';
  const sections = pickSections(artifacts.paperSections);
  const figures = pickFigures(artifacts.figures);
  const refs = pickReferences(artifacts);

  const sectionLines = sections
    .map((s) => `## ${s.key}\n\n${sectionToText(s.key, s.value)}`)
    .join('\n\n');
  const figureLines =
    figures.length > 0
      ? figures.map((c, i) => `### 图 ${i + 1}\n\n${c}`).join('\n\n')
      : '';
  const refLines = refs.length > 0 ? refs.map((r, i) => `${i + 1}. ${r}`).join('\n') : '_（暂无）_';

  return `# ${title}\n\n> 学科: ${discipline}\n\n${sectionLines}\n\n${figureLines}\n\n## 参考文献\n\n${refLines}\n`;
}

/** 简单转义 LaTeX / Markdown 特殊字符（保守处理） */
function escapeTemplate(input: string): string {
  return input
    .replace(/\\/g, '\\textbackslash{}')
    .replace(/&/g, '\\&')
    .replace(/%/g, '\\%')
    .replace(/\$/g, '\\$')
    .replace(/#/g, '\\#')
    .replace(/_/g, '\\_')
    .replace(/\{/g, '\\{')
    .replace(/\}/g, '\\}');
}

/** 全量替换，避免依赖 ES2021 的 replaceAll */
function replaceAll(source: string, token: string, value: string): string {
  return source.split(token).join(value);
}

/** 列出全部支持的模板 */
export function listTemplates(): DraftTemplate[] {
  return [...TEMPLATES];
}

/** 根据模板返回对应下载扩展名 */
export function templateExtension(template: DraftTemplate): string {
  switch (template) {
    case 'ctex':
    case 'ieee':
    case 'journal':
      return 'tex';
    case 'markdown':
      return 'md';
    default:
      return 'txt';
  }
}

/** 渲染时根据模板选择 MIME（下载接口使用） */
export function templateMime(template: DraftTemplate): string {
  switch (template) {
    case 'ctex':
    case 'ieee':
    case 'journal':
      return 'application/x-tex';
    case 'markdown':
      return 'text/markdown';
    default:
      return 'text/plain';
  }
}

File: llm_service/src/agents/draft_renderer.py
"""论文草稿渲染器。

将流水线产生的 artifacts 渲染为不同格式的论文草稿文本：
- ctex      中文双栏论文（ctexart 风格）
- ieee      英文双栏会议（IEEE conference 风格）
- journal   中文期刊
- markdown  通用 Markdown

不依赖任何外部库，纯字符串模板，方便在任何环境运行。
LLM 失败时也能正常渲染（只是内容是占位文字）。
"""

from __future__ import annotations

from typing import Any


# ============================================================================
# 工具函数
# ============================================================================


def _safe(value: Any, default: str = "（未生成）") -> str:
    """安全获取文本字段，None / 空 字符串 / 空 dict 都降级为占位。"""
    if value is None:
        return default
    if isinstance(value, str):
        return value if value.strip() else default
    if isinstance(value, (int, float)):
        return str(value)
    return str(value)


def _section(sections: dict[str, Any], key: str, default: str = "（未生成）") -> str:
    """从 paper_sections 中取某章节文本。"""
    if not sections:
        return default
    return _safe(sections.get(key), default)


def _format_refs(literature: list[dict]) -> str:
    """把文献列表格式化为学术参考文献条目（GB/T 7714 风格简化版）。"""
    if not literature:
        return "（暂无参考文献）"
    lines: list[str] = []
    for i, lit in enumerate(literature, 1):
        title = _safe(lit.get("title"), "[无题名]")
        authors = lit.get("authors") or []
        # 作者格式：最多 3 名，超过用"等"
        if authors:
            if len(authors) <= 3:
                author_str = ", ".join(authors)
            else:
                author_str = ", ".join(authors[:3]) + ", 等"
        else:
            author_str = "佚名"
        year = lit.get("year") or "n.d."
        doi = lit.get("doi")
        url = lit.get("url")
        source = lit.get("source") or ""
        # 构建引用尾部
        tail_parts: list[str] = []
        if doi:
            tail_parts.append(f"DOI: {doi}")
        elif url:
            tail_parts.append(f"URL: {url}")
        # 标注来源类型（非用户上传的才标注检索来源）
        source_tag = ""
        if source and source not in ("unknown", "placeholder", "user_upload", "llm"):
            source_tag = f" [{source}]"
        elif source == "user_upload":
            source_tag = " [用户上传]"
        tail = ", ".join(tail_parts)
        tail_str = f". {tail}" if tail else ""
        lines.append(f"[{i}] {author_str}. {title}[{year}]{source_tag}{tail_str}")
    return "\n".join(lines)


def _format_figure(fig: dict[str, Any], idx: int) -> str:
    """把单个图表规格格式化为可读段落。"""
    fid = _safe(fig.get("id"), f"fig_{idx}")
    ftype = _safe(fig.get("type"), "未指定")
    caption = _safe(fig.get("caption"), "（无说明）")
    data = fig.get("data")
    data_str = "" if data is None else f"  - 数据：{data}"
    code = fig.get("code")
    code_str = "" if not code else f"\n  - 示例代码：\n```python\n{code}\n```"
    return (
        f"图 {idx}（{fid}）\n"
        f"  - 类型：{ftype}\n"
        f"  - 说明：{caption}{data_str}{code_str}"
    )


def _format_figures(figures: list[dict]) -> str:
    if not figures:
        return "（暂无图表）"
    return "\n\n".join(_format_figure(f, i) for i, f in enumerate(figures, 1))


# ============================================================================
# 模板渲染器
# ============================================================================


def render_markdown(artifacts: dict[str, Any]) -> str:
    """通用 Markdown 模板。"""
    title = _safe(artifacts.get("title") or artifacts.get("question"), "未命名论文")
    sections = artifacts.get("paper_sections") or {}
    literature = artifacts.get("literature") or []
    return (
        f"# {title}\n\n"
        f"## 摘要\n{_section(sections, 'abstract')}\n\n"
        f"## 1 引言\n{_section(sections, 'intro', '（未生成）')}\n\n"
        f"## 2 方法\n{_section(sections, 'method')}\n\n"
        f"## 3 结果\n{_section(sections, 'results')}\n\n"
        f"## 4 讨论\n{_section(sections, 'discussion')}\n\n"
        f"## 5 结论\n{_section(sections, 'conclusion')}\n\n"
        f"## 参考文献\n{_format_refs(literature)}\n"
    )


def render_ctex(artifacts: dict[str, Any]) -> str:
    """中文双栏论文（ctexart 风格）。"""
    title = _safe(artifacts.get("title") or artifacts.get("question"), "未命名论文")
    sections = artifacts.get("paper_sections") or {}
    literature = artifacts.get("literature") or []
    figures = artifacts.get("figures") or []
    return (
        r"\documentclass[twocolumn]{ctexart}" + "\n"
        r"\usepackage{graphicx}" + "\n"
        r"\usepackage{amsmath}" + "\n"
        r"\usepackage{cite}" + "\n"
        r"\title{" + _latex_escape(title) + r"}" + "\n"
        r"\author{RAP Agent}" + "\n"
        r"\date{\today}" + "\n"
        r"\begin{document}" + "\n"
        r"\twocolumn[" + "\n"
        r"  \begin{@twocolumnfalse}" + "\n"
        r"    \maketitle" + "\n"
        r"    \begin{abstract}" + "\n"
        f"    {_latex_escape(_section(sections, 'abstract'))}\n"
        r"    \end{abstract}" + "\n"
        r"  \end{@twocolumnfalse}" + "\n"
        r"]" + "\n"
        r"\section{引言}" + f"\n{_latex_escape(_section(sections, 'intro'))}\n\n"
        r"\section{方法}" + f"\n{_latex_escape(_section(sections, 'method'))}\n\n"
        r"\section{结果}" + f"\n{_latex_escape(_section(sections, 'results'))}\n\n"
        r"\section{讨论}" + f"\n{_latex_escape(_section(sections, 'discussion'))}\n\n"
        r"\section{结论}" + f"\n{_latex_escape(_section(sections, 'conclusion'))}\n\n"
        r"\section{图表清单}" + f"\n{_latex_escape(_format_figures(figures))}\n\n"
        r"\section{参考文献}" + f"\n{_latex_escape(_format_refs(literature))}\n\n"
        r"\end{document}" + "\n"
    )


def render_ieee(artifacts: dict[str, Any]) -> str:
    """英文双栏会议（IEEE conference 风格）。"""
    title = _safe(artifacts.get("title") or artifacts.get("question"), "Untitled Paper")
    sections = artifacts.get("paper_sections") or {}
    literature = artifacts.get("literature") or []
    return (
        r"\documentclass[conference]{IEEEtran}" + "\n"
        r"\usepackage{cite}" + "\n"
        r"\usepackage{amsmath,amssymb,amsfonts}" + "\n"
        r"\usepackage{graphicx}" + "\n"
        r"\title{" + _latex_escape(title) + r"}" + "\n"
        r"\author{\IEEEauthorblockN{RAP Agent}" + "\n"
        r"\IEEEauthorblockA{Research Automation Pipeline}}" + "\n"
        r"\begin{document}" + "\n"
        r"\maketitle" + "\n"
        r"\begin{abstract}" + f"\n{_latex_escape(_section(sections, 'abstract'))}\n"
        r"\end{abstract}" + "\n\n"
        r"\section{Introduction}" + f"\n{_latex_escape(_section(sections, 'intro'))}\n\n"
        r"\section{Methodology}" + f"\n{_latex_escape(_section(sections, 'method'))}\n\n"
        r"\section{Results}" + f"\n{_latex_escape(_section(sections, 'results'))}\n\n"
        r"\section{Discussion}" + f"\n{_latex_escape(_section(sections, 'discussion'))}\n\n"
        r"\section{Conclusion}" + f"\n{_latex_escape(_section(sections, 'conclusion'))}\n\n"
        r"\section*{References}" + f"\n{_latex_escape(_format_refs(literature))}\n\n"
        r"\end{document}" + "\n"
    )


def render_journal(artifacts: dict[str, Any]) -> str:
    """中文期刊模板。"""
    title = _safe(artifacts.get("title") or artifacts.get("question"), "未命名论文")
    sections = artifacts.get("paper_sections") or {}
    literature = artifacts.get("literature") or []
    figures = artifacts.get("figures") or []
    return (
        f"# {title}\n\n"
        f"**摘要**：{_section(sections, 'abstract')}\n\n"
        f"**关键词**：{_safe(artifacts.get('keywords'), '科研自动化；论文生成；LLM')}\n\n"
        f"---\n\n"
        f"## 1 引言\n\n{_section(sections, 'intro')}\n\n"
        f"## 2 研究方法\n\n{_section(sections, 'method')}\n\n"
        f"## 3 研究结果\n\n{_section(sections, 'results')}\n\n"
        f"## 4 讨论\n\n{_section(sections, 'discussion')}\n\n"
        f"## 5 结论\n\n{_section(sections, 'conclusion')}\n\n"
        f"## 图表\n\n{_format_figures(figures)}\n\n"
        f"## 参考文献\n\n{_format_refs(literature)}\n"
    )


# ============================================================================
# 主入口
# ============================================================================


_RENDERERS = {
    "ctex": render_ctex,
    "ieee": render_ieee,
    "journal": render_journal,
    "markdown": render_markdown,
}


def render(artifacts: dict[str, Any], template: str = "markdown") -> str:
    """根据 template 渲染 artifacts 为对应格式的论文草稿。

    未知 template 自动降级为 markdown。LLM 未配置或 artifacts 字段缺失时，
    仍会返回包含占位文字的完整草稿，不会抛错。
    """
    template = (template or "markdown").lower()
    renderer = _RENDERERS.get(template, render_markdown)
    try:
        return renderer(artifacts or {})
    except Exception:  # noqa: BLE001 - 渲染层兜底，避免任何异常打断流水线
        # 极端情况下（如 artifacts 数据类型严重异常），降级为 markdown
        return render_markdown(_coerce_safe(artifacts or {}))


def _coerce_safe(artifacts: dict[str, Any]) -> dict[str, Any]:
    """把 artifacts 中的所有关键字段强制转为安全类型，避免渲染时异常。"""
    safe: dict[str, Any] = {}
    for k, v in artifacts.items():
        if isinstance(v, (str, int, float, bool)):
            safe[k] = v
        elif isinstance(v, list):
            safe[k] = v  # 列表内部由 _format_* 自行处理
        elif isinstance(v, dict):
            safe[k] = v
        else:
            safe[k] = str(v) if v is not None else ""
    return safe


def _latex_escape(text: str) -> str:
    """简单 LaTeX 字符转义，避免特殊字符破坏编译。"""
    if not text:
        return ""
    # 只做最常见且安全的转义；不去除中文
    replacements = {
        "\\": r"\textbackslash{}",
        "&": r"\&",
        "%": r"\%",
        "$": r"\$",
        "#": r"\#",
        "_": r"\_",
        "{": r"\{",
        "}": r"\}",
        "~": r"\textasciitilde{}",
        "^": r"\textasciicircum{}",
    }
    out = []
    for ch in text:
        out.append(replacements.get(ch, ch))
    return "".join(out)
